library/src/scala/quoted/Quotes.scala

```diff
@@ -1120,8 +1121,12 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
     /** `TypeTest` that allows testing at runtime in a pattern match if a `Tree` is a `Typed` */
     given TypedTypeTest: TypeTest[Tree, Typed]
 
-    /** Tree representing a type ascription `x: T` in the source code */
-    type Typed <: Term
+    /** Tree representing a type ascription `x: T` in the source code.
+     *
+     *  Also represents a pattern that contains a term `x`.
+     *  Other `: T` patterns use the more general `TypeTree`.
+     */
+    type Typed <: Term & TypeTree
```

Comment:
Is the `TypeTree` here supposed to be `TypedTree`? Also I don't understand why there is both Typed and TypedTree.

Reply:
> Also I don't understand why there is both Typed and TypedTree.

We need to keep `Typed` for backward compatibility. If not we would have been able to only use `TypedTree` and name it `Typed`.

Reply:
> Is the TypeTree here supposed to be TypedTree?

Yes, fixed

Reply:
> We need to keep Typed for backward compatibility.

But could we have `type Typed = TypedTree` instead of it being a subtype?

Reply:
The issue is the `val Type` (and `given TypedMethods`). That one cannot be changed without breaking the API.

Reply:
Could those be deprecated/hidden as `private[scala]` ?

Reply:
We could deprecate the, but we would have no way to add the new API because the signatures would collide.

Reply:
We can always use `@targetName` to avoid signature clashes, but alternatively I think we could use a separate given for the new methods and make the previous one private:
```diff
-    type TypedTree <: Term
+    @deprecated(...) type TypedTree = Typed
...
-    given TypedTreeMethods: TypedTreeMethods
...
+    private[scala] given TypedTreeMethods: TypedTreeMethods
...
+    given TypedMethods: TypedMethods
+    trait TypedTreeMethods:
...
```

Reply:
... or maybe we could keep Typed as is and simply change the signatures of the methods?
```diff
trait TypedModule { this: Typed.type =>
...
-  def apply(expr: Term, tpt: TypeTree): Typed
+  def apply(expr: Tree, tpt: TypeTree): Typed
```
I believe that would be binary and tasty compatible

Reply:
Ok, but it seems this would not work with the `unapply`. Making it return `(Tree, TypeTree)` would at least break source compatibility. It seems it would also break binary compatibility because current user code would be missing a test form `Tree` to `Term`.

Reply:
For the unapply users would need to do
```diff
-          case Typed(expr, tpt) =>
+          case Typed(expr: Term, tpt) =>
```